src/pudl/workspace/resource_cache.py

```diff
@@ -135,6 +137,50 @@ def contains(self, resource: PudlResourceKey) -> bool:
         """Returns True if resource is present in the cache."""
         return self._blob(resource).exists()
 
+class AWSS3Cache(AbstractCache):
+    """Implements file cache backed by AWS S3 bucket."""
+
+    def __init__(self, s3_path: str, **kwargs: Any):
+        """Constructs new cache that stores files in AWS S3 Storage.
+
+        Args:
+            s3_path (str): path to where the data should be stored. This should
+              be in the form of s3://{bucket-name}/{optional-path-prefix}
+        """
+        super().__init__(**kwargs)
+        parsed_url = urlparse(s3_path)
+        if parsed_url.scheme != "s3":
+            raise ValueError(f"s3_path should start with s3:// (found: {s3_path})")
+        self._path_prefix = Path(parsed_url.path)
+        self._s3_client = boto3.resource('s3')
+        self._bucket = self._s3_client.Bucket(parsed_url.netloc)
+
+    def _blob(self, resource: PudlResourceKey) -> Blob:
+        """Retrieve Blob object associated with given resource."""
+        p = (self._path_prefix / resource.get_local_path()).as_posix().lstrip('/')
+        return self._bucket.Object(p)
+
+    def get(self, resource: PudlResourceKey) -> bytes:
+        """Retrieves value associated with given resource."""
+        return self._blob(resource).get()['Body'].read()
+
+    def add(self, resource: PudlResourceKey, value: bytes):
+        """Adds (or updates) resource to the cache with given value."""
+        # TODO return self._blob(resource).upload_from_string(value)
+        return self._blob(resource).put(Body=value)
+
+    def delete(self, resource: PudlResourceKey):
+        """Deletes resource from the cache."""
+        return self._blob(resource).delete()
+
+    def contains(self, resource: PudlResourceKey) -> bool:
+        """Returns True if resource is present in the cache."""
+        try:
+            self._blob(resource).load()
+            return True
+        except:
+            return False
```

Comment:
you maybe want something more like this to be more robust in catching other errors.

```python
try:
    self._blob(resource).load()
except botocore.exceptions.ClientError as e:
    if e.response['Error']['Code'] == "404":
        # The object does not exist.
        return False
    else:
        # Something else has gone wrong.
        raise
else:
    # The object does exist.
    return True
```

Reply:
I was hoping you'd be able to suggest a more robust way of doing this 😂 

I knew we wouldn't want a plain `except` but wasn't sure where to start with something more robust - thanks!